### PR TITLE
UBLOX7PLUS protocol + dynamic configuration

### PR DESCRIPTION
**Pull request by sambas**
Not yet functional (outputting right bytes though) pseudocode of UBLOX7PLUS protocol and dynamic configuration. Just for discussion

### CODE REVIEW

**Comment by digitalentity** [comment]
Cool! Getting rid of hard-coded binary messages is a great improvement that should give more flexibility and might save some space.

**Comment by sambas** [comment]
hmm actually it work now, tested with neo6 and neo7. Some cleaning still required and maybe split message groups like it was before

**Comment by rb1205** [comment]
Mmm strange, I just compiled developement and your branch for NAZE target and yours is _bigger_ than developement by 148 bytes. Are there still some pre-made configuration string lingering somewhere?

**Comment by sambas** [comment]
optimizing comes next

**Comment by sambas** [comment]
I was hoping for more flash space saving, but -8 bytes now on CC3D target. Could shave some bytes with shorter cli protocol string.

File: src/main/fc/cli.c [hunk 1]
```diff
@@ -213,7 +213,7 @@ static const char * const lookupTableAlignment[] = {
 
 #ifdef GPS
 static const char * const lookupTableGPSProvider[] = {
-    "NMEA", "UBLOX", "I2C-NAV", "NAZA"
+    "NMEA", "UBLOX", "I2C-NAV", "NAZA", "UBLOX7PLUS"
```
Comment (digitalentity) [inline src/main/fc/cli.c]: Probably simply UBLOX7 would be enough

File: src/main/io/gps_ublox.c [hunk 2]
```diff
@@ -532,81 +595,80 @@ static bool gpsNewFrameUBLOX(uint8_t data)
...
         break;
 
     case 1: // NAVX5 - skip
-        //ubxTransmitAutoConfigCommands(ubloxInit_NAVX5, sizeof(ubloxInit_NAVX5));
         gpsState.autoConfigStep++;
         break;
 
     case 2: // Disable NMEA messages
-        ubxTransmitAutoConfigCommands(ubloxInit_MSG_NMEA, sizeof(ubloxInit_MSG_NMEA));
+        configureMSG(MSG_CLASS_NMEA, MSG_NMEA_GGA, 0);
```
Comment (digitalentity) [inline src/main/io/gps_ublox.c:618]: Would it be better to split this bunch of messages over different `autoConfigStep` values?

**Comment by stronnag** [comment]
flight tested on OMNIBUS and SPRF3 (fwiw).

File: src/main/io/gps_ublox.c (hunk 2 above)
Comment (sambas) [reply src/main/io/gps_ublox.c:618]: Yeah, I can do that. One message per step?

**Comment by sambas** [comment]
stronnag, thanks for testing. Did you try with ublox7 protocol?

File: src/main/fc/cli.c (hunk 1 above)
Comment (sambas) [reply src/main/fc/cli.c]: changing it

**Comment by stronnag** [comment]
Yes, I should have been explicit:
````
# get gps
gps_provider = UBLOX7PLUS
````

**Comment by stronnag** [comment]
So much fun, I flew it twice (with ublox7plus)

image

File: src/main/io/gps_ublox.c (hunk 2 above)
Comment (digitalentity) [reply src/main/io/gps_ublox.c:618]: Actually, it's probably ok as-is. Previously we were sending bulk messages no problem - this approach shouldn't be any different.

**Comment by giacomo892** [comment]
I'm trying this on NAZA but it doesn't compile if I don'#define GPS_PROTO_UBLOX_NEO7PLUS in conjunction with #define GPS_PROTO_UBLOX.  (undefined reference to `gpsHandleUBLOX')

What are the benefits using UBLOX7 vs standard UBLOX beside the update rate that now is 10-12.5Hz?

**Comment by stronnag** [comment]
Benefits include 10hz operation, better bandwidth utilisation (PVT stanza v. multiple stanzas for same data). So potentially better data consistency.

And owner happiness because they are using the full potential of the device, even if there are few practical benefits.

**Comment by giacomo892** [comment]
Thanks. So as far as I see this is a rewrite of ublox decoder but it is still depended on old ublox decoder. 
It seems to work with mine M7N,M8N and NAZE correctly on the bench.

**Comment by sambas** [comment]
just extension, not rewrite

**Comment by giacomo892** [comment]
Ok. I'll try this as soon 1.7.1 will be official and see what it has to offer :)

**Pull request merged**